### PR TITLE
Fix java bindings

### PR DESCRIPTION
**Pull request by vvviiimmm**
1. `vm_execute` input data was not handled properly. An additional `inputData` parameter has been added to `EvmcVm`'s `execute` function to avoid pointer arithmetics in java
2. `ReleaseByteArrayElements` was called on a pointer that is not derived from `GetByteArrayElements` causing segfaults
3. `get_block_hash_fn` typo

Also, it was quite challenging to encode `evmc_message` into ByteBuffer and then decode the result. I'd suggest capturing that logic into java classes so the callers won't have to work with memory buffers directly.

### CODE REVIEW

**Comment by axic** [comment]
Thanks! It seems to fail on CI, can you please check?

> Also, it was quite challenging to encode evmc_message into ByteBuffer and then decode the result. I'd suggest capturing that logic into java classes so the callers won't have to work with memory buffers directly.

Absolutely, we did some work on it but didn't had the time to finish. See #542 and #555.

File: bindings/java/c/host.c [hunk 1]
```diff
@@ -369,7 +363,7 @@ static struct evmc_tx_context get_tx_context_fn(struct evmc_host_context* contex
 
 static evmc_bytes32 get_block_hash_fn(struct evmc_host_context* context, int64_t number)
 {
-    char java_method_name[] = "get_code_hash";
+    char java_method_name[] = "get_block_hash_fn";
```
Comment (axic) [inline bindings/java/c/host.c:366]: Oops. Shows the lack of units tests for Java.

File: bindings/java/c/evmc-vm.c [hunk 2]
```diff
@@ -112,6 +113,10 @@ JNIEXPORT jobject JNICALL Java_org_ethereum_evmc_EvmcVm_execute(JNIEnv* jenv,
     struct evmc_result* result =
         (struct evmc_result*)(*jenv)->GetDirectBufferAddress(jenv, jresult);
     assert(result != NULL);
+    size_t input_data_size;
+    const uint8_t* input_data = GetDirectBuffer(jenv, jinputdata, &input_data_size);
+    char* input_data_addr = (char*)msg + offsetof(struct evmc_message, input_data);
+    memcpy(input_data_addr, &input_data, sizeof(uint8_t*));
```
Comment (axic) [inline bindings/java/c/evmc-vm.c:119]: Why does it need to use this convoluted approach of memory copying, as opposed to just `msg.input_data = GetDirectBuffer(jenv, jinputdata, &input_data_size);` ?

File: bindings/java/c/host.c [hunk 3]
```diff
@@ -276,8 +272,6 @@ static size_t copy_code_fn(struct evmc_host_context* context,
             memcpy(buffer_data, code + code_offset, length);
     }
 
-    (*jenv)->ReleaseByteArrayElements(jenv, jaddress, (jbyte*)address, 0);
```
Comment (axic) [inline bindings/java/c/host.c:279]: > ReleaseByteArrayElements was called on a pointer that is not derived from GetByteArrayElements causing segfaults

These helpers rely on `CopyDataToJava` which uses `NewByteArray/SetByteArrayRegion`. Are you saying those cannot be freed with it?

File: bindings/java/c/host.c (hunk 1 above)
Comment (axic) [reply bindings/java/c/host.c:366]: Extracted this into #653.

File: bindings/java/c/evmc-vm.c (hunk 2 above)
Comment (axic) [reply bindings/java/c/evmc-vm.c:119]: Actually should be `msg.input_data = GetDirectBuffer(jenv, jinputdata, &msg.input_size);` as the size setting is missing right now.

File: bindings/java/c/evmc-vm.c [hunk 4]
```diff
@@ -97,7 +97,8 @@ JNIEXPORT jobject JNICALL Java_org_ethereum_evmc_EvmcVm_execute(JNIEnv* jenv,
                                                                 jobject jcontext,
                                                                 jint jrev,
                                                                 jobject jmsg,
-                                                                jobject jcode)
+                                                                jobject jcode,
+                                                                jobject jinputdata)
```
Comment (axic) [inline bindings/java/c/evmc-vm.c:101]:
```suggestion
                                                                jobject jinput)
```

Would name it like this if there are no symbol clashes.

File: bindings/java/c/host.c (hunk 3 above)
Comment (vvviiimmm) [reply bindings/java/c/host.c:279]: Can't say for sure, but isn't `NewByteArray` creating a local ref that will be garbage collected? https://docs.oracle.com/javase/10/docs/specs/jni/functions.html#local-references

File: bindings/java/java/src/main/java/org/ethereum/evmc/EvmcVm.java [hunk 5]
```diff
@@ -144,16 +144,21 @@ public String version() {
    * <p>This is a mandatory method and MUST NOT be set to NULL.
    */
   private static native ByteBuffer execute(
-      ByteBuffer nativeVm, HostContext context, int rev, ByteBuffer msg, ByteBuffer code);
+      ByteBuffer nativeVm,
+      HostContext context,
+      int rev,
+      ByteBuffer msg,
+      ByteBuffer code,
+      ByteBuffer inputdata);
```
Comment (axic) [inline bindings/java/java/src/main/java/org/ethereum/evmc/EvmcVm.java:152]: Same here.

File: bindings/java/java/src/test/java/org/ethereum/evmc/EvmcTest.java [hunk 6]
```diff
@@ -230,16 +255,21 @@ void testExecute_EVMC_CREATE() throws Exception {
...
-      char[] inputData = "hello w\0".toCharArray();
+      char[] inputDataPlaceholder = {};
+      byte[] inputDataBytes = {0x68, 0x65, 0x6c, 0x6c, 0x6f};
+      ByteBuffer inputDataByteBuffer =
+          ByteBuffer.wrap(inputDataBytes).order(ByteOrder.nativeOrder());
       long gas = 200000;
       int depth = 0;
       ByteBuffer msg =
-          new TestMessage(kind, sender, recipient, value, inputData, gas, depth).toByteBuffer();
+          new TestMessage(kind, sender, recipient, value, inputDataPlaceholder, gas, depth)
```
Comment (axic) [inline bindings/java/java/src/test/java/org/ethereum/evmc/EvmcTest.java:265]: I was wondering what this `TestMessage` is and why it needs the `inputDataPlaceholder`.

It turns out `vm.execute` still gets this entire data structure with `input_data`/`input_length`.

This is a bad idea having shadow data which is not checked. Probably the better option is removing those from the `msg` structure, which also means it needs to be reassambled in the binding.

**Comment by chfast** [comment]
@vvviiimmm Are you using EVMC in Java?

**Comment by axic** [comment]
With #706 Java support was dropped -- closing this.

**Pull request closed without merging**